File: couples/python/prefs.py
import sys, random

class Prefs:
    def __init__(self, quantity):
        self.values = [0 for i in range(quantity ** 2)]
        self.quantity = quantity

    def get_value(self, first_index, second_index):
        return self.values[first_index * self.quantity + second_index]

    def set_value(self, first_index, second_index, value):
        self.values[first_index * self.quantity + second_index] = value

    def init_with_value(self, value):
        for i in range(self.quantity ** 2):
            self.quantity[i] = value

    def init_with_zeros(self):
        self.init_with_value(0)

    def is_unique(self, value, first_index, second_index):
        for i in range(0, second_index):
            if self.get_value(first_index, i) == value:
                return False
        return True

    def init_with_random_values(self):
        for i in range(self.quantity):
            for j in range(self.quantity):
                value = 0
                while (True):
                    value = random.randint(0, self.quantity - 1)
                    if self.is_unique(value, i, j):
                        break
                self.set_value(i, j, value)

    def show(self):
        for i in range(self.quantity):
            for j in range(self.quantity):
                print(self.get_value(i, j), end = " ")
            print()

File: couples/java/src/zeionara/couples/Prefs.java
package zeionara.couples;

import java.util.concurrent.ThreadLocalRandom;

public class Prefs implements DataStore, IntMatrix{
	private int[] values;
	private int quantity;
	
	public Prefs(int quantity){
		this.quantity = quantity;
		values = new int[quantity * quantity];
		init_with_zeros();
	}

	public int get_quantity(){
		return quantity;
	}

	public int get_value(int first_index, int second_index){
		return values[first_index * quantity + second_index];
	}

	public void set_value(int first_index, int second_index, int value){
		values[first_index * quantity + second_index] = value;
	}

	public void init_with_value(int value){
		for (int i = 0; i < quantity * quantity; i++){
			values[i] = value;
		}
	}

	private boolean is_unique(int first_index, int second_index, int value){
		for (int i = 0; i < second_index; i++){
			if (get_value(first_index, i) == value){
				return false;
			}
		}
		return true;
	}


	public void init_with_random_values(){
		for (int i = 0; i < quantity; i++){
			for (int j = 0; j < quantity; j++){
				int random_value = 0;
				do {
					random_value = ThreadLocalRandom.current().nextInt(0, quantity);
				} while (! is_unique(i, j, random_value));
				set_value(i, j, random_value);
			}
		}
	}


	public void init_with_zeros(){
		init_with_value(0);
	}

	public void show(){
		for (int i = 0; i < quantity; i++){
			for (int j = 0; j < quantity; j++){
				System.out.print(get_value(i, j) + " ");
			}
			System.out.println();
		}
	}
}

